Do not accidentally override `n` operand when generating `e` and `c` operands

devversion opened the pull request:
Recently support for the `e` and `c` operands has been added. Without
realizing, the `n` operand is accidentally overwritten when the operand
variables are initialized. The `n` variable is used as a temporary
accumulator but this variable might be accessed later in the plural
rules, so it's incorrect to change its value.

@papandreou Sorry! Looks like the change wasn't as good as I thought 😄 I realized
that using `n` as a temporary variable could break the use of `n` later in the plural rules.

I'm not sure what you think of this solution. It seems a bit odd to me that the helper variable
is introduced manually through AST. I guess this could be more simpler if the whole function
body is used and evaluated.

devversion commented:
FWIW: I considered using a simple RegExp replace as with the other operands. This could work but makes it more tricky, and return unexpected (IMO) results if the input string is bad. e.g 10e10e10 will return 10 while I'd expect zero due to an invalid number.

Though while testing this it looked like using inputs like "5e4" fail anyway because `parseInt` parses the number without respecting the `e`. It just returns 5.

papandreou commented:
> Though while testing this it looked like using inputs like "5e4" fail anyway because `parseInt` parses the number without respecting the `e`. It just returns 5. I have an idea how we can fix that.

We could use `parseFloat` instead?

```js
> parseFloat('5e4')
50000
```

devversion commented:
Yeah, exactly. I was also thinking about throwing if the parsed number becomes `NaN`

devversion commented:
@papandreou I've updated the PR. Here are a few things:

1. I decided to use the `Number` function instead of `parseFloat`. This ensures that we don't accidentally parse parts of an incorrect input. e.g. `pluralFn('10abcde');`. The `Number` fn will return `NaN` in such cases.
2. I've reworked a little how the `n` operand is computed. This allows users to actually pass numbers with the exponent/mantisse shorthand. e.g. `plurlaFn('10e4');`. This didn't work before.
3. I've added a check for `isNaN`. An error is thrown if a number cannot be parsed. This is better than incorrectly computing operands with the various RegExp's for the operands.

I think this should be ready for your review. Thanks!

papandreou commented:
The throwing makes it semver-major I guess. Right now you kinda get away with passing in bogus:

```js
> require('cldr').extractPluralRuleFunction('fr')('foobar')
'other'
```

But I do agree that it's better to reject invalid input.

papandreou reviewed (approved):
LGTM -- and thanks! 🥇

papandreou commented:
Released in 7.0.0.
